File: algorithms/23_elixir/graphs/bfs.ex
defmodule BFS do
  def bfs(graph, start) do
    bfs_loop(graph, :queue.from_list([start]), MapSet.new([start]), [])
  end
  defp bfs_loop(_, queue, _, acc) when :queue.is_empty(queue), do: Enum.reverse(acc)
  defp bfs_loop(graph, queue, visited, acc) do
    {{:value, node}, rest} = :queue.out(queue)
    neighbors = Map.get(graph, node, [])
    {new_queue, new_visited} = Enum.reduce(neighbors, {rest, visited}, fn n, {q, v} ->
      if MapSet.member?(v, n), do: {q, v}, else: {:queue.in(n, q), MapSet.put(v, n)}
    end)
    bfs_loop(graph, new_queue, new_visited, [node | acc])
  end
end

File: algorithms/08_r/graphs/bfs.R
bfs <- function(adj_list, start_node) {
  visited <- c()
  queue <- c(start_node)
  order <- c()
  
  while (length(queue) > 0) {
    node <- queue[1]
    queue <- queue[-1]
    
    if (!(node %in% visited)) {
      visited <- c(visited, node)
      order <- c(order, node)
      neighbors <- adj_list[[as.character(node)]]
      queue <- c(queue, neighbors)
    }
  }
  return(order)
}
